File: js/quick_sort.js
function quickSort(array, first, last) {
    if (last > first) {
        let pivotIndex = partition(array, first, last);
        quickSort(array, first, pivotIndex - 1);
        quickSort(array, pivotIndex + 1, last);
    }
}

function partition(array, first, last) {
    let pivot = array[first];
    let low = first + 1;
    let high = last;

    while (high > low) {
        while (low <= high && array[low] <= pivot) {
            low++;
        }

        while (low <= high && array[high] > pivot) {
            high--;
        }

        if (high > low) {
            let temp = array[high];
            array[high] = array[low];
            array[low] = temp;
        }
    }

    while (high > first && array[high] >= pivot) {
        high--;
    }

    if (pivot > array[high]) {
        array[first] = array[high];
        array[high] = pivot;
        return high;
    } else {
        return first;
    }
}

function testQuickSort() {
    output.innerHTML += "<h3>testing quick sort in javascript...</h3>";
    let unsortedArray = getRandArray(10, 1);
    output.innerHTML += `<p>unsorted array -> ${unsortedArray}</p>`;
    quickSort(unsortedArray, 0, unsortedArray.length - 1);
    output.innerHTML += `<p>sorted array -> ${unsortedArray}</p>`;
}

testQuickSort();

File: python/quick_sort.py
def quick_sort(list, first, last):
    if last > first:
        pivot_index = partition(list, first, last)
        quick_sort(list, first, pivot_index - 1)
        quick_sort(list, pivot_index + 1, last)



def partition(list, first, last):
    pivot = list[first]
    low = first + 1
    high = last

    while high > low:
        while low <= high and list[low] <= pivot:
            low += 1

        while low <= high and list[high] > pivot:
            high -= 1

        if high > low:
            temp = list[high]
            list[high] = list[low]
            list[low] = temp
    
    while high > first and list[high] >= pivot:
        high -= 1

    if pivot > list[high]:
        list[first] = list[high]
        list[high] = pivot
        return high
    else:
        return first

print("testing quick sort in python...")

unsortedArray = [4, 1, 5, 2, 8, 3, 9, 7, 6]
print("unsorted array -> " + str(unsortedArray))
quick_sort(unsortedArray, 0, len(unsortedArray) - 1)
print("sorted array -> " + str(unsortedArray))